File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        longest = 0
        l = 0
        sett = set()

        for r in range(len(s)):
            # while window is not valid
            while s[r] in sett:
                sett.remove(s[l])
                l += 1
            longest = max(longest, r-l+1)
            sett.add(s[r])
        return longest

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> hashmap = new HashMap<>();
        int l = 0;
        int r = 0;
        int longestWindow = 0;
        while (r < s.length()) {
            while (hashmap.containsKey(s.charAt(r)) && hashmap.get(s.charAt(r)) >= l) {
                l = hashmap.get(s.charAt(r)) + 1;
            }
            hashmap.put(s.charAt(r), r);
            longestWindow = Math.max(longestWindow, r-l+1);
            r++;
        }
        return longestWindow;
    }
}